From: Dzmitry Sankouski <dsankouski@gmail.com>
Date: Sun, 23 Feb 2025 19:45:08 +0300
Subject: arm64: dts: qcom: sdm845-starqltechn: improve support

arm64: dts: qcom: sdm845: enable gmu

Leave gmu enabled, because it's only probed when
GPU is.

Reviewed-by: Konrad Dybcio <konrad.dybcio@oss.qualcomm.com>
Signed-off-by: Dzmitry Sankouski <dsankouski@gmail.com>

arm64: dts: qcom: sdm845-starqltechn: remove wifi

Starqltechn has broadcom chip for wifi, so sdm845 wifi part
can be disabled.

Fixes: d711b22eee55 ("arm64: dts: qcom: starqltechn: add initial device tree for starqltechn")
Reviewed-by: Konrad Dybcio <konrad.dybcio@linaro.org>
Signed-off-by: Dzmitry Sankouski <dsankouski@gmail.com>

arm64: dts: qcom: sdm845-starqltechn: fix usb regulator mistake

Usb regulator was wrongly pointed to vreg_l1a_0p875.
However, on starqltechn it's powered from vreg_l5a_0p8.

Fixes: d711b22eee55 ("arm64: dts: qcom: starqltechn: add initial device tree for starqltechn")
Reviewed-by: Konrad Dybcio <konrad.dybcio@oss.qualcomm.com>
Signed-off-by: Dzmitry Sankouski <dsankouski@gmail.com>

arm64: dts: qcom: sdm845-starqltechn: refactor node order

Fixes: d711b22eee55 ("arm64: dts: qcom: starqltechn: add initial device tree for starqltechn")
Signed-off-by: Dzmitry Sankouski <dsankouski@gmail.com>

arm64: dts: qcom: sdm845-starqltechn: remove excess reserved gpios

Starqltechn has 2 reserved gpio ranges <27 4>, <85 4>.
<27 4> is spi for eSE(embedded Secure Element).
<85 4> is spi for fingerprint.

Remove excess reserved gpio regions.

Fixes: d711b22eee55 ("arm64: dts: qcom: starqltechn: add initial device tree for starqltechn")
Reviewed-by: Konrad Dybcio <konrad.dybcio@oss.qualcomm.com>
Signed-off-by: Dzmitry Sankouski <dsankouski@gmail.com>

arm64: dts: qcom: sdm845-starqltechn: add gpio keys

Add support for phone buttons.

Reviewed-by: Konrad Dybcio <konrad.dybcio@oss.qualcomm.com>
Signed-off-by: Dzmitry Sankouski <dsankouski@gmail.com>

arm64: dts: qcom: sdm845-starqltechn: add max77705 PMIC

Add support for max77705 MFD device. Supported sub-devices:
 charger, fuelgauge, haptic, led

Reviewed-by: Konrad Dybcio <konrad.dybcio@oss.qualcomm.com>
Signed-off-by: Dzmitry Sankouski <dsankouski@gmail.com>

arm64: dts: qcom: sdm845-starqltechn: add display PMIC

Add support for s2dos05 display / touchscreen PMIC

Reviewed-by: Konrad Dybcio <konrad.dybcio@oss.qualcomm.com>
Signed-off-by: Dzmitry Sankouski <dsankouski@gmail.com>

arm64: dts: qcom: sdm845-starqltechn: add touchscreen support

Add support for samsung,s6sy761 touchscreen.

Reviewed-by: Konrad Dybcio <konrad.dybcio@oss.qualcomm.com>
Signed-off-by: Dzmitry Sankouski <dsankouski@gmail.com>

arm64: dts: qcom: sdm845-starqltechn: add initial sound support

Add support for sound (headphones and mics only)
Also redefine slpi reserved memory, because adsp_mem overlaps with
slpi_mem inherited from sdm845.dtsi.

Signed-off-by: Dzmitry Sankouski <dsankouski@gmail.com>

arm64: dts: qcom: sdm845-starqltechn: add graphics support

Add support for gpu and panel.

Reviewed-by: Konrad Dybcio <konrad.dybcio@oss.qualcomm.com>
Signed-off-by: Dzmitry Sankouski <dsankouski@gmail.com>

arm64: dts: qcom: sdm845-starqltechn: add modem support

Add support for modem and ipa(IP Accelerator).
Add spss reserved memory node.

Reviewed-by: Konrad Dybcio <konrad.dybcio@oss.qualcomm.com>
Signed-off-by: Dzmitry Sankouski <dsankouski@gmail.com>

arm64: dts: qcom: sdm845-starqltechn: add slpi support

Add slpi node to support sensors

arm64: dts: qcom: sdm845-starqltechn: add framebuffer reserved memory

arm64: dts: qcom: sdm845-starqltechn: disable fb regulators

Disable framebuffer regulators, to allow panel to switch off screen
completely

arm64: dts: qcom: sdm845-starqltechn: add more audio support
---
 .../boot/dts/qcom/sdm845-samsung-starqltechn.dts   | 90 ++++++++++++++++++++++
 1 file changed, 90 insertions(+)

diff --git a/arch/arm64/boot/dts/qcom/sdm845-samsung-starqltechn.dts b/arch/arm64/boot/dts/qcom/sdm845-samsung-starqltechn.dts
index 978ff50..4b84e9c 100644
--- a/arch/arm64/boot/dts/qcom/sdm845-samsung-starqltechn.dts
+++ b/arch/arm64/boot/dts/qcom/sdm845-samsung-starqltechn.dts
@@ -14,6 +14,7 @@
 #include <dt-bindings/sound/qcom,q6afe.h>
 #include <dt-bindings/sound/qcom,q6asm.h>
 #include <dt-bindings/sound/qcom,wcd934x.h>
+#include <dt-bindings/sound/qcom,q6voice.h>
 
 #include "sdm845.dtsi"
 #include "pm8998.dtsi"
@@ -122,6 +123,11 @@ spss_mem: spss@97700000 {
 			no-map;
 		};
 
+		fb_mem: memory@95800000 {
+			reg = <0 0x95800000 0 0x500000>;
+			no-map;
+		};
+
 		adsp_mem: memory@97800000 {
 			reg = <0 0x97800000 0 0x2000000>;
 			no-map;
@@ -138,6 +144,34 @@ rmtfs_mem: rmtfs-mem@fde00000 {
 		};
 	};
 
+	i2c18 {
+		compatible = "i2c-gpio";
+		sda-gpios = <&tlmm 62 GPIO_ACTIVE_HIGH>;
+		scl-gpios = <&tlmm 63 GPIO_ACTIVE_HIGH>;
+		i2c-gpio,delay-us = <2>;
+		pinctrl-0 = <&i2c18_sda_state &i2c18_scl_state>;
+		pinctrl-names = "default";
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		max98512_codec: max98512@39 {
+			compatible = "maxim,max98512";
+			#sound-dai-cells = <0>;
+			reg = <0x39>;
+			maxim,spk-gain = <0x5>;
+			maxim,spk-gain-rcv = <0x1>;
+			maxim,digital-gain-rcv = <0x34>;
+			maxim,sysclk = <0xbb8000>;
+			maxim,vmon_slot = <0x1>;
+			maxim,platform_info = <0x1 0x1006 0x10001062 0x1ff1020b 0x0 0x33>;
+			maxim,boost_step = <0xf 0xf 0xf 0xf 0xf 0xf 0xf 0xf 0xf 0xf 0xf 0xf 0xf 0xf 0x2 0x0>;
+			maxim,registers-of-amp = <0x1f 0x0 0x21 0xc0 0x2d 0x14 0x2f 0x0>;
+			maxim,sub_reg = <0x3a>;
+			maxim,mono_stereo = <0x1>;
+			interleave_mode = <0x1>;
+		};
+	};
+
 	i2c21 {
 		compatible = "i2c-gpio";
 		sda-gpios = <&tlmm 127 GPIO_ACTIVE_HIGH>;
@@ -753,6 +787,28 @@ cpu {
 		};
 	};
 
+	voicemmode1-dai-link {
+		link-name = "VoiceMMode1";
+		cpu {
+			sound-dai = <&q6voicedai VOICEMMODE1>;
+		};
+	};
+
+	speaker-dai-link {
+		link-name = "Speaker Playback";
+		codec {
+			sound-dai = <&max98512_codec>;
+		};
+
+		cpu {
+			sound-dai = <&q6afedai QUATERNARY_MI2S_RX>;
+		};
+
+		platform {
+			sound-dai = <&q6routing>;
+		};
+	};
+
 	slim-dai-link {
 		link-name = "SLIM Playback 1";
 
@@ -872,6 +928,18 @@ dai@5 {
 	};
 };
 
+&q6cvp {
+	status = "okay";
+};
+
+&q6cvs {
+	status = "okay";
+};
+
+&q6mvm {
+	status = "okay";
+};
+
 &wcd9340 {
 	reset-gpios = <&tlmm 64 GPIO_ACTIVE_HIGH>;
 	vdd-buck-supply = <&vreg_s4a_1p8>;
@@ -898,6 +966,14 @@ &ipa {
 	status = "okay";
 };
 
+&slpi_pas {
+	firmware-name = "qcom/sdm845/starqltechn/slpi.mbn";
+	status = "okay";
+
+//	vdd-cx-supply = <&vdd-l1-l27-supply>;
+//	vdd-cx-supply = <&vdd-l4-l5-supply>;
+};
+
 &usb_1 {
 	status = "okay";
 };
@@ -1024,6 +1100,20 @@ sd_card_det_n_state: sd-card-det-n-state {
 		bias-pull-up;
 	};
 
+	i2c18_sda_state: i2c18-sda-state {
+		pins = "gpio62";
+		function = "gpio";
+		drive-strength = <2>;
+		bias-disable;
+	};
+
+	i2c18_scl_state: i2c18-scl-state {
+		pins = "gpio63";
+		function = "gpio";
+		drive-strength = <2>;
+		bias-disable;
+	};
+
 	touch_irq_state: touch-irq-state {
 		pins = "gpio120";
 		function = "gpio";
